provide matrix_symbolic_sparse class

dimpase opened the pull request:
sparse matrices over SR are broken without this

This is needed in particular to specialize echelonize(), cf #35653

dimpase committed:
provide matrix_symbolic_sparse class

This is needed in particulat to specialize echelonize(), cf #35653

dimpase committed:
correct spacing to please linter

mkoeppe commented on src/sage/matrix/matrix_symbolic_sparse.pyx:
```diff
@@ -0,0 +1,1034 @@
+        TESTS::
+
+            sage: m = matrix([[cos(pi/5), sin(pi/5)], [-sin(pi/5), cos(pi/5)]])
+            sage: m.echelonize(); m
```
Is the whole thing copy-pasted from `matrix_symbolic_dense`? 

Note that the copy-pasted doctests from here to the end of the file do not use `sparse=True`, so they won't test your new class.

dimpase replied:
there are indirect doctests with sparse=True elsewhere which test this class

But it looks I forgot a commit somewhere. Thanks for catching this.

dimpase replied:
it's mostly copypasted, of course. It looks as if at some point I just got bored with adding `sparse=True`, and forgot about the rest of the doctests.

dimpase replied:
OK, this is fixed - and sparse symbolic vectors had to be added

dimpase committed:
sparse=True in all examples/tests, fix for jordan_form()

dimpase committed:
vector_symbolic_sparse: needed for matrix_symbolic_sparse

without it, simplify_full() and canonicalize_radical() do not work

also fixed docs by adding few */modules/*sparse* missing in the index

dimpase committed:
a bit different error message due to the new class

mkoeppe reviewed (approved):
Otherwise LGTM.
